### PR TITLE
feat(must-gather): Added must-gather scripts

### PR DESCRIPTION
**Pull request by vimalk78**
Thie PR adds `must-gather` scripts for kepler-operator

Sample execution on single node CRC cluster:
```
$ oc adm must-gather --image=$(oc -n openshift-operators get deployment.apps/kepler-operator-controller-manager -o jsonpath='{.spec.template.spec.containers[?(@.name == "manager")].image}')  -- /usr/bin/gather 
[must-gather      ] OUT Using must-gather plug-in image: quay.io/vimalkum/kepler-operator:0.0.0-must-gather
When opening a support case, bugzilla, or issue please include the following summary data along with any other requested information:
ClusterID: 1f7b621c-7e7a-4ca8-81d5-f49ad9d9acb6
ClusterVersion: Stable at "4.13.12"
ClusterOperators:
	clusteroperator/cloud-credential is missing
	clusteroperator/cluster-autoscaler is missing
	clusteroperator/insights is missing
	clusteroperator/monitoring is missing
	clusteroperator/storage is missing


[must-gather      ] OUT namespace/openshift-must-gather-t75p4 created
[must-gather      ] OUT clusterrolebinding.rbac.authorization.k8s.io/must-gather-llgwf created
[must-gather      ] OUT pod for plug-in image quay.io/vimalkum/kepler-operator:0.0.0-must-gather created
[must-gather-mjjlz] POD 2023-10-03T09:38:03.616361057Z must-gather logs are located at: '/must-gather/gather-debug.log'
[must-gather-mjjlz] POD 2023-10-03T09:38:03.616361057Z powermon must-gather started...
[must-gather-mjjlz] POD 2023-10-03T09:38:03.629416554Z 2023-10-03 09:38:03 getting kepler instance
[must-gather-mjjlz] POD 2023-10-03T09:38:05.108103126Z 2023-10-03 09:38:05 getting openshift-kepler-operator events
[must-gather-mjjlz] POD 2023-10-03T09:38:05.697635288Z 2023-10-03 09:38:05 getting kepler exporter daemonset
[must-gather-mjjlz] POD 2023-10-03T09:38:06.905154905Z 2023-10-03 09:38:06 getting kepler exporter config map
[must-gather-mjjlz] POD 2023-10-03T09:38:07.653189798Z 2023-10-03 09:38:07 getting kepler exporter service account
[must-gather-mjjlz] POD 2023-10-03T09:38:08.162324862Z 2023-10-03 09:38:08 getting kepler exporter service account
[must-gather-mjjlz] POD 2023-10-03T09:38:08.785281522Z 2023-10-03 09:38:08 running gather script for kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:08.864821711Z 2023-10-03 09:38:08 collecting pod yaml for kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:09.663837981Z 2023-10-03 09:38:09 collecting information from "cpuid" from kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:10.371721220Z 2023-10-03 09:38:10 collecting environment variables from kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:11.223976399Z 2023-10-03 09:38:11 collecting kernel version from kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:12.969152179Z 2023-10-03 09:38:12 collecting ebpf information from kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:14.204599060Z 2023-10-03 09:38:14 collecting logs from kepler pod: kepler-exporter-ds-h26gp
[must-gather-mjjlz] POD 2023-10-03T09:38:14.821072700Z 2023-10-03 09:38:14 running gather script for olm
[must-gather-mjjlz] POD 2023-10-03T09:38:14.914602992Z 2023-10-03 09:38:14 collecting olm info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:18.226907411Z 2023-10-03 09:38:18 collecting olm summary
[must-gather-mjjlz] POD 2023-10-03T09:38:19.342571734Z 2023-10-03 09:38:19 getting kepler-operator info
[must-gather-mjjlz] POD 2023-10-03T09:38:19.380959531Z 2023-10-03 09:38:19 collecting subscription info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:19.839823768Z 2023-10-03 09:38:19 collecting catalogsource info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:20.203048340Z 2023-10-03 09:38:20 collecting installplan info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:20.930806572Z 2023-10-03 09:38:20 collecting CSV for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:24.948783859Z 2023-10-03 09:38:24 collecting deployment info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:29.524788569Z 2023-10-03 09:38:29 collecting pod info for kepler-operator
[must-gather-mjjlz] POD 2023-10-03T09:38:48.171054954Z powermon must-gather completed
[must-gather-mjjlz] OUT waiting for gather to complete
[must-gather-mjjlz] OUT downloading gather output
[must-gather-mjjlz] OUT receiving incremental file list
[must-gather-mjjlz] OUT ./
[must-gather-mjjlz] OUT gather-debug.log
[must-gather-mjjlz] OUT kepler-exporter-cm.yaml
[must-gather-mjjlz] OUT kepler-exporter-ds.yaml
[must-gather-mjjlz] OUT kepler-exporter-sa.yaml
[must-gather-mjjlz] OUT kepler-exporter-scc.yaml
[must-gather-mjjlz] OUT kepler.yaml
[must-gather-mjjlz] OUT openshift-kepler-operator_events
[must-gather-mjjlz] OUT kepler-info/
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/ebpf-info
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/env-variables
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/kepler-pod.yaml
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/kepler.log
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/kernel-info
[must-gather-mjjlz] OUT kepler-info/kepler-exporter-ds-h26gp/node-cpuid-info
[must-gather-mjjlz] OUT kepler-operator-info/
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator-catalogsource.yaml
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator-csv.yaml
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator-deployment.yaml
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator-installplan.yaml
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator-subscription.yaml
[must-gather-mjjlz] OUT kepler-operator-info/kepler-operator.yaml
[must-gather-mjjlz] OUT kepler-operator-info/summary.txt
[must-gather-mjjlz] OUT olm-info/
[must-gather-mjjlz] OUT olm-info/olm-reources.yaml
[must-gather-mjjlz] OUT olm-info/summary.txt
[must-gather-mjjlz] OUT 
[must-gather-mjjlz] OUT sent 469 bytes  received 137,494 bytes  30,658.44 bytes/sec
[must-gather-mjjlz] OUT total size is 592,732  speedup is 4.30
[must-gather      ] OUT namespace/openshift-must-gather-t75p4 deleted
[must-gather      ] OUT clusterrolebinding.rbac.authorization.k8s.io/must-gather-llgwf deleted


Reprinting Cluster State:
When opening a support case, bugzilla, or issue please include the following summary data along with any other requested information:
ClusterID: 1f7b621c-7e7a-4ca8-81d5-f49ad9d9acb6
ClusterVersion: Stable at "4.13.12"
ClusterOperators:
	clusteroperator/cloud-credential is missing
	clusteroperator/cluster-autoscaler is missing
	clusteroperator/insights is missing
	clusteroperator/monitoring is missing
	clusteroperator/storage is missing


```

### CODE REVIEW

File: must-gather/gather [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+print_usage() {
+    echo -e "Usage: /usr/bin/gather <gather-collection-path>"
+}
+
+case ${1} in
+  --help | -h )
+	print_usage
+	exit 1
+esac
```
Comment (sthaha) [inline must-gather/gather:24]: Could we move this to `main` so that all execution walk-through can start from `main` ?

File: must-gather/common [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+
+run() {
+  length=$(($#-1))
+  cmd=${@:1:$length}
+  gather_file="${@: -1}"
+
+  log "$cmd"
+  $cmd >> $gather_file  2>> $LOGFILE_PATH
```
Comment (sthaha) [inline must-gather/common:33]: could you please run shellcheck on all scripts? I think it should catch a lot of bash gotchas that we overlook during a review. 

suggestion: May be this is a good PR to add shellcheck to our CI as well. Or have that added before this PR is merged since this script will be used in production.

File: must-gather/gather-olm-info [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
```
Comment (sthaha) [inline must-gather/gather-olm-info:2]: all shell script should have `set -eu -o pipefail` to ensure that the script is robust

File: must-gather/gather-olm-info [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+source ${SCRIPT_DIR}/common
+
+BASE_COLLECTION_PATH=$1
+LOGFILE_PATH="${BASE_COLLECTION_PATH}/${LOGFILE_NAME}"
+
+OLM_INFO_DIR=$BASE_COLLECTION_PATH/olm-info
+mkdir -p $OLM_INFO_DIR
```
Comment (sthaha) [inline must-gather/gather-olm-info:10]: can we move all side-effects of running as script to start from main?

**Review by sthaha — CHANGES_REQUESTED** [review]

File: must-gather/common (hunk 2 above)
Comment (sthaha) [reply must-gather/common:33]: I have added https://github.com/sustainable-computing-io/kepler-operator/pull/262 which adds shellcheck and  fixes shellcheck issues in existing scripts.

File: must-gather/common [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+export KEPLER_NS="openshift-kepler-operator"
+export LOGFILE_NAME="gather-debug.log"
```
Comment (sthaha) [inline must-gather/common:4]: To help with maintenance (and to make debugging easier) it is better to avoid side-effects as much as possible  when sourcing files. Ideally, it is better if the file only included variables (esp  globals) which is only used by the functions in that file.  Hence I would recommend moving `KEPLER_NS` to individual files. Yes we are repeating ourselves but I believe that is the lesser evil :) 

If you notice the use of `LOGFILE_NAME`, is unused in this file while a `LOGFILE_PATH` is required by the functions and that must be defined elsewhere.  I think it may be easier to encapsulate this by requiring the caller to invoke and `init_log` function that sets the `LOGFILE_PATH` and makes it readonly soon after.

File: Dockerfile [hunk 6]
```diff
@@ -12,6 +12,7 @@ COPY go.sum go.sum
 RUN go mod download
 
 # Copy the go source
+COPY must-gather/ must-gather/
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply Dockerfile:15]: this is usually how multi stage builds are done. COPY happens in first stage. but i am open to changing it.

File: must-gather/common [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+get_timestamp(){
+  date '+%Y-%m-%d %H:%M:%S'
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply must-gather/common:7]: this is for capturing timestamps in `gather-debug.log` 
must-gather doesnt add timestamps, what we see on the console is `gather` container logs which could be different from `gather-debug.log`.

File: must-gather/common [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply must-gather/common:1]: done

File: must-gather/common [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+export KEPLER_NS="openshift-kepler-operator"
+export LOGFILE_NAME="gather-debug.log"
+
+get_timestamp(){
+  date '+%Y-%m-%d %H:%M:%S'
+}
+
+log() {
+  echo "$(get_timestamp) ${*}"
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply must-gather/common:11]: changed

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply Dockerfile:15]: AFAIK, we only need to copy what is for each stage required, so moving the `copy` to line 40 should work.

File: must-gather/common (hunk 7 above)
[the remark this answers is not in the document]
Comment (sthaha) [reply must-gather/common:7]: I see ... thanks for the explanation :)

File: must-gather/common (hunk 2 above)
Comment (vimalk78) [reply must-gather/common]: dome shellcheck, and fixed

File: must-gather/gather-olm-info (hunk 4 above)
Comment (vimalk78) [reply must-gather/gather-olm-info]: done

File: must-gather/gather-olm-info (hunk 3 above)
Comment (vimalk78) [reply must-gather/gather-olm-info:19]: done

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (vimalk78) [reply Dockerfile:15]: first stage is where all copying is done, this helps with caching.

[the remark this answers is not in the document]
Comment (sthaha) [reply Dockerfile:15]: Actually, we don't need to cache. The rule of thumb is that COPY to the layer where it is used. 
I am going to resolve this in the interest of time :)

**Review by sthaha — APPROVED** [review]
Looks good to me buy I will get @vprashar2929 merge if after the validation is done.

File: must-gather/gather-kepler-operator-info [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  KEPLER_OPERATOR_INFO_DIR="$BASE_COLLECTION_PATH/kepler-operator-info"
+  mkdir -p "$KEPLER_OPERATOR_INFO_DIR"
+
+  get_subscription
+  get_catalogsource
+  get_install_plan
+  get_csv
+  get_kepler_operator_deployment_info
+  get_kepler_operator_pod_info
+  get_summary
```
Comment (sthaha) [inline must-gather/gather-kepler-operator-info:89]: should we add || true to every step ?

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (vimalk78) [reply Dockerfile:15]: it is about podman/docker cache

File: must-gather/gather-kepler-operator-info (hunk 10 above)
Comment (sthaha) [reply must-gather/gather-kepler-operator-info:89]: `set -e` should have stopped the script.

Comment (vimalk78) [reply must-gather/gather-kepler-operator-info:89]: good that i didn't add it then :)

File: must-gather/gather [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+  echo "powermon must-gather started..."
+  get_kepler_instance
+  get_kepler_events
+  get_kepler_daemon_set
+  get_kepler_config_map
+  get_kepler_sa
+  get_kepler_scc
+  gather_kepler_exporter_info
+  gather_olm_info
```
Comment (vprashar2929) [inline must-gather/gather:107]: shouldn't we begin with capturing olm and operator info?

File: must-gather/gather [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+source "${SCRIPT_DIR}"/common
+
+print_usage() {
+    echo -e "Usage: /usr/bin/gather <gather-collection-path>"
+}
+
+get_kepler_instance() {
+  log "getting kepler instance"
+  run oc get keplers.kepler.system.sustainable.computing.io kepler -oyaml "$BASE_COLLECTION_PATH/kepler.yaml"
```
Comment (vprashar2929) [inline must-gather/gather:30]: In case of Kepler instance is not created the execution gets terminated and the rest of the steps are skipped. I also believe we should add `|| true` at every step

File: must-gather/gather-kepler-operator-info (hunk 10 above)
Comment (sthaha) [reply must-gather/gather-kepler-operator-info:89]: I will verify this but we should always ensure that these steps break (exit with non-zero value)  due to the nature of bash. E.g. say there is a typo in any of the command above, we should be able to find that outduring development we would comment out `# || true` in the optional steps. `get_catalog_source` should be optional since that is will only be present if you `run bundle`

File: must-gather/gather (hunk 11 above)
Comment (sthaha) [reply must-gather/gather:107]: In the test scenario, I installed only the operator (imagine the operator installation failed) and ran the must gather. 
```
❯ less after-install/quay-io-sthaha-kepler-operator-sha256-df656da3b9f7a0c4310f8a44cb7b11eb790a7be43794e67c1343e934953f6478/gather-debug.log
2023-10-05 01:23:52 getting kepler instance
2023-10-05 01:23:52 oc get keplers.kepler.system.sustainable.computing.io kepler -oyaml
Error from server (NotFound): keplers.kepler.system.sustainable.computing.io "kepler" not found
```

And thats all it captures

File: must-gather/gather [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+
+source "${SCRIPT_DIR}"/common
+
+print_usage() {
+    echo -e "Usage: /usr/bin/gather <gather-collection-path>"
+}
+
+get_kepler_instance() {
```
Comment (sthaha) [inline must-gather/gather:28]: I think we should get all `keplers` instead of just the `one` until we have a webhook that prevents creating any other kepler.

File: must-gather/gather-kepler-operator-info [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+  run oc -n "$KEPLER_OPERATOR_NS" get deployment "$KEPLER_OPERATOR_DEPLOY" -oyaml "$KEPLER_OPERATOR_INFO_DIR/kepler-operator-deployment.yaml"
+}
+
+get_kepler_operator_pod_info() {
+  log "collecting pod info for kepler-operator" 
+  run oc -n "$KEPLER_OPERATOR_NS" get pod -l control-plane=controller-manager -oyaml "$KEPLER_OPERATOR_INFO_DIR/kepler-operator.yaml"
+}
+
+get_summary() {
+  run oc -n "$KEPLER_OPERATOR_NS" get catalogsource kepler-operator-catalog -owide "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
```
Comment (sthaha) [inline must-gather/gather-kepler-operator-info:59]: In the scenario where operator is installed from `community-catalog`, `get_summary` breaks here  and the script terminates execution
```
2023-10-05 01:24:48 oc -n openshift-operators get subscription -l operators.coreos.com/kepler-operator.openshift-operators= -oyaml
2023-10-05 01:24:48 collecting catalogsource info for kepler-operator
2023-10-05 01:24:48 oc -n openshift-operators get catalogsource kepler-operator-catalog -oyaml
Error from server (NotFound): catalogsources.operators.coreos.com "kepler-operator-catalog" not found
~
```

File: must-gather/gather-kepler-operator-info [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  echo -e "\n" >> "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+
+  run oc -n "$KEPLER_OPERATOR_NS" get subscription -owide "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+  echo -e "\n" >> "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+
+  run oc -n "$KEPLER_OPERATOR_NS" get installplan -owide "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+  echo -e "\n" >> "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+
+  run oc -n "$KEPLER_OPERATOR_NS" get csv -owide  "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
+  echo -e "\n" >> "$KEPLER_OPERATOR_INFO_DIR/summary.txt"
```
Comment (sthaha) [inline must-gather/gather-kepler-operator-info:69]: suggestion: how about we add `\n` in `run` function itself?

File: must-gather/gather [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+  BASE_COLLECTION_PATH="${1:-/must-gather}"
+  export LOGFILE_PATH="${BASE_COLLECTION_PATH}/${LOGFILE_NAME}"
+
+  mkdir -p "${BASE_COLLECTION_PATH}"
+  cd "${BASE_COLLECTION_PATH}"
+  echo -e "must-gather logs are located at: '${LOGFILE_PATH}'"
+
+  mkdir -p "${BASE_COLLECTION_PATH}/cache-dir"
+  export KUBECACHEDIR=${BASE_COLLECTION_PATH}/cache-dir
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply must-gather/gather:97]: the cache-dir is deleted at the end.

File: must-gather/gather (hunk 11 above)
Comment (vimalk78) [reply must-gather/gather:107]: the script is not stateful.

File: must-gather/gather (hunk 16 above)
[the remark this answers is not in the document]
Comment (vimalk78) [reply must-gather/gather:97]: updated

File: must-gather/gather (hunk 11 above)
Comment (sthaha) [reply must-gather/gather:107]: let me rephrase, in the scenario where the operator fails to install, i.e. no `kepler` CRD in cluster, what information should we capture?

File: must-gather/gather-kepler-operator-info (hunk 10 above)
Comment (vimalk78) [reply must-gather/gather-kepler-operator-info:89]: if we have to deal with so much complexity, i would rather have must-gather written as a go program. `oc adm must-gather` does not care if it is a script or a binary.

File: must-gather/gather (hunk 11 above)
Comment (vimalk78) [reply must-gather/gather:107]: if script executes all its steps, we will capture olm info.

**Comment by vimalk78** [comment]
script runs to completion

File: must-gather/gather (hunk 13 above)
Comment (vimalk78) [reply must-gather/gather:28]: any kepler not named `kepler` is ignored by operator.  
but changed script to capture all keplers

Comment (sthaha) [reply must-gather/gather:28]: Actually, it doesn't get ignored, the operator does run a reconcile and set the status to invalid.

**Review by sthaha — APPROVED** [review]
